src/interface/data_collection.hpp

```diff
@@ -105,6 +120,28 @@ class DataCollection {
   }
 
  private:
+  template <class U>
+  std::string GetKey(const std::string &stage_label, const std::shared_ptr<U> &in) {
+    if constexpr (std::is_same_v<U, MeshData<Real>>) {
+      std::string key = stage_label + "_part-" + std::to_string(in->partition);
+      if (in->grid.type == GridType::two_level_composite)
+        key = key + "_gmg-" + std::to_string(in->grid.logical_level);
+      return key;
+    } else {
+      return stage_label;
+    }
+  }
+  std::string GetKey(const std::string &stage_label, std::optional<int> partition_id,
+                     std::optional<int> gmg_level) {
+    std::string key = stage_label;
+    if (partition_id) key = key + "_part-" + std::to_string(*partition_id);
+    if (gmg_level) key = key + "_gmg-" + std::to_string(*gmg_level);
```

Comment:
Should there be a debug_require that only partition_id or gmg_level or neither is passed?

Reply:
no, the GMG levels are also partitioned. Although to be fair, partitioning has always been broken for multigrid, i.e. it only works when all blocks are included in the partition. This is on my list of things to fix.